internal/mesh/internal/controllers/implicitdestinations/index.go

```diff
@@ -0,0 +1,237 @@
+// Cache: reverse CTP => WI[source]
+var ctpBySourceWorkloadIdentityIndex = indexers.RefOrIDIndex(
+	"ctp-by-source-workload-identity",
+	func(ctp *types.DecodedComputedTrafficPermissions) []*pbresource.Reference {
+		// We ignore wildcards for this index.
+		exact, _, _ := getSourceWorkloadIdentitiesFromCTP(ctp)
+		return maps.Values(exact)
+	},
+)
+
+const ctpByWildcardSourceIndexName = "ctp-by-wildcard-source"
+
+func ctpByWildcardSourceIndexCreator(globalDefaultAllow bool) *index.Index {
+	return indexers.DecodedMultiIndexer(
+		ctpByWildcardSourceIndexName,
+		index.SingleValueFromArgs(func(tn tenantedName) ([]byte, error) {
+			return indexFromTenantedName(tn), nil
+		}),
+		func(r *types.DecodedComputedTrafficPermissions) (bool, [][]byte, error) {
+			var vals [][]byte
+
+			if r.Data.IsDefault && globalDefaultAllow {
+				// Literally everything can reach it.
+				vals = append(vals, indexFromTenantedName(tenantedName{
+					Partition: storage.Wildcard,
+					Namespace: storage.Wildcard,
+					Name:      storage.Wildcard,
+				}))
+				return true, vals, nil
+			}
+
+			for _, perm := range r.Data.AllowPermissions {
+				for _, src := range perm.Sources {
+					srcType := determineSourceType(src)
+					if srcType != sourceTypeLocal {
+						// Partition / Peer / SamenessGroup are mututally exclusive.
+						continue // Ignore these for now.
+					}
+					// It is assumed that src.Partition != "" at this point.
+
+					if src.IdentityName != "" {
+						// exact
+						continue
+					} else if src.Namespace != "" {
+						// wildcard name
+						vals = append(vals, indexFromTenantedName(tenantedName{
+							Partition: src.Partition,
+							Namespace: src.Namespace,
+							Name:      storage.Wildcard,
+						}))
+					} else {
+						// wildcard name+ns
+						vals = append(vals, indexFromTenantedName(tenantedName{
+							Partition: src.Partition,
+							Namespace: storage.Wildcard,
+							Name:      storage.Wildcard,
+						}))
+					}
+				}
+			}
+
+			return true, vals, nil
+		},
+	)
+}
+
+type tenantedName struct {
+	Partition string
+	Namespace string
+	Name      string
+}
+
+func indexFromTenantedName(tn tenantedName) []byte {
+	var b index.Builder
+	b.String(tn.Partition)
+	b.String(tn.Namespace)
+	b.String(tn.Name)
+	return b.Bytes()
+}
+
+// Cache: reverse CR => SVC[backend]
+var computedRoutesByBackendServiceIndex = indexers.RefOrIDIndex(
+	"computed-routes-by-backend-service",
+	func(cr *types.DecodedComputedRoutes) []*pbresource.Reference {
+		return getBackendServiceRefsFromComputedRoutes(cr)
+	},
+)
+
+func getWorkloadIdentitiesFromService(svc *types.DecodedService) []*pbresource.Reference {
+	ids := catalog.GetBoundIdentities(svc.Resource)
+
+	out := make([]*pbresource.Reference, 0, len(ids))
+	for _, id := range ids {
+		out = append(out, &pbresource.Reference{
+			Type:    pbauth.WorkloadIdentityType,
+			Name:    id,
+			Tenancy: svc.Id.Tenancy,
+		})
+	}
+	return out
+}
+
+func getSourceWorkloadIdentitiesFromCTP(
```

Comment:
It's probably worth adding some unit tests for the helper functions in this file since there is quite a bit of logic in them.